flash shell and fs shell: Add read/erase/write test commands

nordicjm opened the pull request:
Adds commands which can be used for timing flash device tasks and
    outputting a rough speed.

    Adds an optional shell command for testing the speed of the flash
    device and file system, which outputs a rough speed.

Fixes #54098

nordicjm committed:
flash: flash_shell: Add flash read/write/erase speed test commands

Adds commands which can be used for timing flash device tasks and
outputting a rough speed.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
fs: fs_shell: Add erase/write test

Adds an optional shell command for testing the speed of the flash
device and file system, which outputs a rough speed.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

Laczen commented:
-1 from my side.

While it would be good to perform some speed testing on flash read/write/erase (this should be a selection of a random area and size and limited to e.g. 32 trials), the idea of using this to find out problems in a communication is bad.

As a result of (mis)using this test a user might result in premature end of flash life.

Instead of trying to find out if communication is a problem by looking at the total of communication and flash write/read/erase it is better to eliminate flash from the equation by providing a stub flash interface.

nordicjm commented:
> As a result of (mis)using this test a user might result in premature end of flash life.

Nonsense, who is going to run 100,000 tests on a single sector? Computers with SSDs have had speed tests ran on them for decades now.

de-nordic commented on drivers/flash/flash_shell.c:
```diff
@@ -255,6 +254,246 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
+	return result;
+}
+
+static int cmd_erase_write_test(const struct shell *sh, size_t argc, char *argv[])
```
I do not think this one is needed. There is usually some distance between erase and write in code, so this does not profile any valid sequence.

Testing operations by type is fine as it allows to profile them and for example profile how much improvement you get from implementing page vs segment erase in memories, or profile improvements or how much changes to drivers or API cost us in performance.
Also this can be used to get flash API busy, even if ops happen to flash RAM driver, and at the same time check how the other parts of system behave or are impacted.

nordicjm replied:
I used this one to check the difference with what the file system one is doing since that is doing erasing and writing, so it gives a like-for-like rough view of the filesystem overhead

Laczen commented:
> > As a result of (mis)using this test a user might result in premature end of flash life.
> 
> Nonsense, who is going to run 100,000 tests on a single sector? Computers with SSDs have had speed tests ran on them for decades now.

SSD's have wear-levelling, raw flash not. 

Doing this test on a fs might be acceptable because they should have wear-levelling (but then again fatfs does not), on raw flash it is unacceptable.

de-nordic reviewed (changes requested):
I can see some use in testing flash access times.
Some comments left.

de-nordic commented:
> 
> Doing this test on a fs might be acceptable because they should have wear-levelling (but then again fatfs does not), on raw flash it is unacceptable.

We are killing a lot of flash during development and testing. I would not consider this feature to be something you want to have in released software but when you are developing or improving driver, this may be simplified way of testing how does your changes affect performance.

Laczen commented:
> > Doing this test on a fs might be acceptable because they should have wear-levelling (but then again fatfs does not), on raw flash it is unacceptable.
> 
> We are killing a lot of flash during development and testing. I would not consider this feature to be something you want to have in released software but when you are developing or improving driver, this may be simplified way of testing how does your changes affect performance.

I see added value in a flash performance test as this could indeed simplify developing or improving drivers, but this test does not provide a flash performance test.

nordicjm commented on drivers/flash/flash_shell.c:
```diff
@@ -255,6 +254,246 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
+	}
+
+	size = strtoul(argv[2], NULL, 16);
+	repeat = strtoul(argv[3], NULL, 16);
```
Fixed

nordicjm commented on subsys/fs/shell.c:
```diff
@@ -477,6 +478,128 @@ static int cmd_write(const struct shell *shell, size_t argc, char **argv)
 
+#ifdef CONFIG_FILE_SYSTEM_SHELL_TEST_COMMANDS
+const static uint8_t speed_types[][3] = { "B", "KB", "MB", "GB" };
+const static uint32_t speed_divisor = 1024;
```
Fixed

nordicjm commented on drivers/flash/flash_shell.c:
```diff
@@ -255,6 +254,246 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
+
+	result = 0;
+
+	while (repeat--) {
+		start_tick = k_uptime_get();
+		result = flash_read(flash_dev, addr, test_arr, size);
+		loop_tick = k_uptime_delta(&start_tick);
+
+		if (result) {
+			shell_error(sh, "Read failed: %d", result);
+			break;
+		}
```
Yes missed that, fixed.

nordicjm commented on drivers/flash/flash_shell.c:
```diff
@@ -255,6 +254,246 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
+	result = 0;
+
+	while (repeat--) {
+		start_tick = k_uptime_get();
```
Fixed

carlescufi commented:
> I see added value in a flash performance test as this could indeed simplify developing or improving drivers, but this test does not provide a flash performance test.

It provides a throughput test @Laczen. Isn't that a performance test?

Laczen commented:
> > I see added value in a flash performance test as this could indeed simplify developing or improving drivers, but this test does not provide a flash performance test.
> 
> It provides a throughput test @Laczen. Isn't that a performance test?

A flash performance test should eliminate the communication from the test,
A communication troughput test should eliminate the flash from the test (could be done by making a flash test/stub device),

I see added value in both, testing them at the same time leaves you with nothing but doubts on communication and flash.

nordicjm commented:
> A flash performance test should eliminate the communication from the test,

What's the point in that then? If a performance test tells you the speed of the memory itself, why would anyone run a test? Go and read the datasheet. The point in a performance test gives you the actual level of performance for a piece of hardware that you can expect on a device. No-one would run a CPU/GPU/HDD/SSD benchmark utility on a PC if it just gave you the number from a datasheet, it would be a pointless test.

Laczen commented:
> > A flash performance test should eliminate the communication from the test,
> 
> What's the point in that then? If a performance test tells you the speed of the memory itself, why would anyone run a test? Go and read the datasheet. The point in a performance test gives you the actual level of performance for a piece of hardware that you can expect on a device. No-one would run a CPU/GPU/HDD/SSD benchmark utility on a PC if it just gave you the number from a datasheet, it would be a pointless test.

A flash performance test could e.g. be used to test the flash driver, you can't read any flash driver errors in a datasheet.

This comment illustrates the fact that this test is only for communication debugging, this should not be performed on real flash but on a flash stub.

nordicjm committed:
fs: fs_shell: Add read test

Adds an optional shell command for testing the speed of the flash
device and file system, which outputs a rough speed for reading a
file.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
This was used internally (with the new read function) to diagnose an issue with read speed of a memory card and used to adjust configuration until a required minimum performance could be achieved. So with that I declare this performance test metric to be very successful for what it was designed for

Laczen commented:
> This was used internally (with the new read function) to diagnose an issue with read speed of a memory card and used to adjust configuration until a required minimum performance could be achieved. So with that I declare this performance test metric to be very successful for what it was designed for

Please elaborate, how did you distinguish between a problem with the communication or the memory card? As you said before it can't have been the memory card (read the datasheet, remember), so it was the communication, this should have been detected with a stub write.

nordicjm commented:
> Please elaborate, how did you distinguish between a problem with the communication or the memory card? As you said before it can't have been the memory card (read the datasheet, remember), so it was the communication, this should have been detected with a stub write.

Class U1 memory card, minimum 10MiBps speed getting 146KiBps read speed, so clearly it was communication, it was the DMA buffer size, changing from the default of 8 to 128 improved read speed to 531KiBps, so no, it has absolutely nothing to do with a stub driver, having a stub driver would have done absolutely nothing to help with this problem or investigation and

Laczen commented:
> > Please elaborate, how did you distinguish between a problem with the communication or the memory card? As you said before it can't have been the memory card (read the datasheet, remember), so it was the communication, this should have been detected with a stub write.
> 
> Class U1 memory card, minimum 10MiBps speed getting 146KiBps read speed, so clearly it was communication, it was the DMA buffer size, changing from the default of 8 to 128 improved read speed to 531KiBps, so no, it has absolutely nothing to do with a stub driver, having a stub driver would have done absolutely nothing to help with this problem or investigation and would have served no purpose.

DMA of what ? On one hand you seem to be using this test to determine a problem with the communication layer when updating flash, is it the same problem with the card, or is it something else... When a user uses this test he/she should know where the problem is.

Laczen commented:
After going through the complete PR it seems like I was put on the wrong track by the issue that generated this PR. The PR is a local test and is not influenced by any communication to the shell. As such this is a good performance test (although I would prefer not to see any flash erases).

I would like to apologize to @nordicjm and should have studied the PR more carefully.
